File: 4 Kyu/4 Kyu - Multiplying numbers as strings.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_DIGIT 1000
#define NUM(x) ((x) - '0')

void reverse (char *src) {

  int tmp, i = 0, j = strlen (src) - 1;

  while (i < j) {
    tmp = src[i];
    src[i] = src[j];
    src[j] = tmp;
    i++, j--;
  }
}
char *multiply(char *a, char *b) {

  int lena = strlen (a), lenb = strlen(b), ans[MAX_DIGIT] = {0};
  int i,j, size;
  char arra[lena + 1], arrb[lenb + 1];

  if ((lena == 1 && a[0] == '0')|| (lenb == 1 && b[0] == '0'))
      return "0";

  strcpy (arra,a), strcpy (arrb,b);
  reverse (arra), reverse (arrb);

  for (i = 0; i < lena; ++i) {
      for (j = 0; j < lenb; ++j) {
          int num = NUM (arra[i]) * NUM (arrb[j]);
          ans[i + j] += num;
          ans[i + j + 1] = ans[i + j + 1] + ans[i + j] / 10;
          ans[i + j] %= 10;
      }
  }

  size = i + j;
  char *result = calloc (size + 1,sizeof(char));
  while (ans[size] == 0) size--;

  i = 0;
  do {
      result[i++] = ans[size] + '0';
  } while (size-->0);

    return result;
}

File: 4 Kyu/4 Kyu - Multiplying numbers as strings.rs
fn reverse (src: &str) -> Vec<u8> {
    src.bytes().rev().collect::<Vec<u8>>()
}
fn multiply(a: &str, b: &str) -> String {

    if a == "0" || b == "0" { return "0".to_string() }

    let a = reverse (a);
    let b = reverse (b);
    let mut os:String = String::new();
    let mut ans = [0; 1000];
    let mut size = 0;

    for i in 0..a.len() {
        for j in 0..b.len() {
            ans[i + j] += (a[i] - 0x30) * (b[j] - 0x30);
            ans[i + j + 1] = ans[i + j + 1] + ans[i + j] / 10;
            ans[i + j] %= 10;
            size = i + j + 1;
        }
    }
    while ans[size] == 0 { size -= 1 }

    loop {
        os.push ((ans[size] + 0x30) as char);
        if size == 0 { return os }
        size -= 1;
    }
}
